pysph/sph/surface_tension.py

```diff
@@ -19,6 +19,198 @@
 from math import sqrt
 
 
+class SurfaceForceAdami(Equation):
+    def initialize(self, d_au, d_av, d_idx):
+        d_au[d_idx] = 0.0
+        d_av[d_idx] = 0.0
+
+    def loop(self, d_au, d_av, d_idx, d_m, DWIJ, d_pi00, d_pi01, d_pi10,
+             d_pi11, s_pi00, s_pi01, s_pi10, s_pi11, d_V, s_V, s_idx):
+        s2 = s_V[s_idx]*s_V[s_idx]
+        f1 = (d_pi00[d_idx]/(d_V[d_idx]*d_V[d_idx])+s_pi00[s_idx]/s2)
+        f2 = (d_pi01[d_idx]/(d_V[d_idx]*d_V[d_idx])+s_pi01[s_idx]/s2)
+        f3 = (d_pi10[d_idx]/(d_V[d_idx]*d_V[d_idx])+s_pi10[s_idx]/s2)
+        f4 = (d_pi11[d_idx]/(d_V[d_idx]*d_V[d_idx])+s_pi11[s_idx]/s2)
+        d_au[d_idx] += (DWIJ[0]*f1 + DWIJ[1]*f2)/d_m[d_idx]
+        d_av[d_idx] += (DWIJ[0]*f3 + DWIJ[1]*f4)/d_m[d_idx]
+
+
+class ConstructStressMatrix(Equation):
+
+    def __init__(self, dest, sources, sigma, d=2):
+        self.sigma = sigma
+        self.d = d
+        super(ConstructStressMatrix, self).__init__(dest, sources)
+
+    def initialize(self, d_pi00, d_pi01, d_pi10, d_pi11, d_cx, d_cy, d_idx,
+                   d_N):
+        mod_gradc2 = d_cx[d_idx]*d_cx[d_idx] + d_cy[d_idx]*d_cy[d_idx]
+        mod_gradc = sqrt(mod_gradc2)
+        d_N[d_idx] = 0.0
+        if mod_gradc > 1e-14:
+            factor = self.sigma/mod_gradc
+            d_pi00[d_idx] = (-d_cx[d_idx]*d_cx[d_idx] +
+                             (mod_gradc2)/self.d)*factor
+            d_pi01[d_idx] = -factor*d_cx[d_idx]*d_cy[d_idx]
+            d_pi10[d_idx] = -factor*d_cx[d_idx]*d_cy[d_idx]
+            d_pi11[d_idx] = (-d_cy[d_idx]*d_cy[d_idx] +
+                             (mod_gradc2)/self.d)*factor
+            d_N[d_idx] = 1.0
+        else:
+            d_pi00[d_idx] = 0.0
+            d_pi01[d_idx] = 0.0
+            d_pi10[d_idx] = 0.0
+            d_pi11[d_idx] = 0.0
```

Comment:
Same here, only 2d?

Reply:
Same as above.

Reply:
This has been fixed to work for 3D also. Please take a look